### PR TITLE
[SDL-0238] Test scripts for "Keyboard Enhancements" feature

### PR DESCRIPTION
**Pull request by dboltovskyi**
ATF Test Scripts to check: 
 - [FORDTCN-10940](https://adc.luxoft.com/jira/browse/FORDTCN-10939)
 - [FORDTCN-10940](https://adc.luxoft.com/jira/browse/FORDTCN-10940)
 - [FORDTCN-10940](https://adc.luxoft.com/jira/browse/FORDTCN-10941)

This PR is **[ready]** for review.

### Summary
New scripts for the feature

### ATF version
[feature/7.1.0_version_bump](https://github.com/smartdevicelink/sdl_atf/tree/feature/7.1.0_version_bump)

### Changelog
 - Added new scripts for the feature
 - Added new test set for the feature

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: test_scripts/API/KeyboardEnhancements/Capabilities/002_HMI_provides_valid_value_for_one_parameter_in_KeyboardCapabilities.lua [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+end
+
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+runner.Step("Register App", common.registerApp)
+
+runner.Title("Test")
+for tc, data in pairs(tcs) do
```
Comment (GetmanetsIrina) [inline test_scripts/API/KeyboardEnhancements/Capabilities/002_HMI_provides_valid_value_for_one_parameter_in_KeyboardCapabilities.lua:48]: Let's use `spairs` from` utils.lua` for sequential steps execution

File: test_scripts/API/KeyboardEnhancements/Capabilities/005_HMI_provides_invalid_value_for_one_parameter_in_KeyboardCapabilities.lua [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+end
+
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+runner.Step("Register App", common.registerApp)
+
+runner.Title("Test")
+for tc, data in pairs(tcs) do
```
Comment (GetmanetsIrina) [inline test_scripts/API/KeyboardEnhancements/Capabilities/005_HMI_provides_invalid_value_for_one_parameter_in_KeyboardCapabilities.lua:58]: Let's use spairs from utils.lua for sequential steps execution

File: test_scripts/API/KeyboardEnhancements/OnKeyboardInput/001_App_receives_OnKeyboardInput.lua [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+runner.Step("Clean environment", common.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", common.start)
+runner.Step("Register App", common.registerApp)
+runner.Step("PTU", common.policyTableUpdate, { ptUpd })
+runner.Step("Activate App", common.activateApp)
+
+runner.Title("Test")
+runner.Step("HMI sends OnSCU", common.sendOnSCU)
+runner.Step("App sends SetGP", common.sendSetGP, { sgpParams, common.result.success })
+runner.Step("HMI sends OnKI", common.sendOnKI, { { data = "k", event = "INPUT_KEY_MASK_ENABLED" } })
```
Comment (GetmanetsIrina) [inline test_scripts/API/KeyboardEnhancements/OnKeyboardInput/001_App_receives_OnKeyboardInput.lua:49]: I believe that in this case notification should be sent without `data`

**Review by GetmanetsIrina — CHANGES_REQUESTED** [review]

**Comment by GetmanetsIrina** [comment]
The review is still in progress

**Comment by GetmanetsIrina** [comment]
Review is finished. There are no additional comments.

**Comment by dboltovskyi** [comment]
@GetmanetsIrina Please notice all comments have been addressed in a12e250

**Review by GetmanetsIrina — APPROVED** [review]

**Comment by GetmanetsIrina** [comment]
@dboltovskyi, thank you for the updates.

File: test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+      keyboardLayout = "NUMERIC",
+      customizeKeys = { "^" } -- not supported special character
+    }
+  }
+  local dataToHMI = common.cloneTable(sgpParams)
+  dataToHMI.appID = common.getHMIAppId()
+  local cid = common.getMobileSession():SendRPC("SetGlobalProperties", sgpParams)
+  common.getHMIConnection():ExpectRequest("UI.SetGlobalProperties", dataToHMI)
+  :Do(function(_, data)
+      common.getHMIConnection():SendError(data.id, data.method, "WARNINGS", msg)
```
Comment (VjKlepikov) [inline test_scripts/API/KeyboardEnhancements/SetGlobalProperties/008_App_defines_not_supported_special_character_in_customizeKeys.lua:38]:
```suggestion
      common.getHMIConnection():SendResponse(data.id, data.method, "WARNINGS", {})
```
I propose to use the successful response type for the result code: "WARNINGS".

File: test_scripts/API/KeyboardEnhancements/Capabilities/005_HMI_provides_invalid_value_for_one_parameter_in_KeyboardCapabilities.lua [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+--  - Respond with DATA_NOT_AVAILABLE, success:false to App
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Local Variables ]]
+local keyboardLayouts = { "QWERTY", "QWERTZ", "AZERTY", "NUMERIC" }
+
+local tcs = {
+  [01] = { maskInputCharactersSupported = "false" },
```
Comment (VjKlepikov) [inline test_scripts/API/KeyboardEnhancements/Capabilities/005_HMI_provides_invalid_value_for_one_parameter_in_KeyboardCapabilities.lua:22]:
```suggestion
  [01] = { maskInputCharactersSupported = "false" }, --invalid type
```
I propose to add a description for all mentioned cases  e.g. : upper out of bound or invalid type

File: test_scripts/API/KeyboardEnhancements/Capabilities/003_HMI_doesnt_provide_KeyboardCapabilities.lua [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+----------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local common = require('test_scripts/API/KeyboardEnhancements/common')
+
+--[[ Local Variables ]]
+local dispCaps = common.getDispCaps()
+dispCaps.systemCapability.displayCapabilities[1].windowCapabilities[1].keyboardCapabilities = nil
+
+--[[ Local Functions ]]
+local function check(_, data)
```
Comment (VjKlepikov) [inline test_scripts/API/KeyboardEnhancements/Capabilities/003_HMI_doesnt_provide_KeyboardCapabilities.lua:23]:
```suggestion
local function checkKeyboardCapabilities(_, data)
```
I propose to rename the function

**Review by VjKlepikov — APPROVED** [review]

File: test_scripts/API/KeyboardEnhancements/Capabilities/003_HMI_doesnt_provide_KeyboardCapabilities.lua (hunk 6 above)
Comment (dboltovskyi) [reply test_scripts/API/KeyboardEnhancements/Capabilities/003_HMI_doesnt_provide_KeyboardCapabilities.lua:23]: I would agree if there are many "check" functions. However we have just only one in some scripts.

**Comment by dboltovskyi** [comment]
@VjKlepikov Please notice all comments have been addressed in 3f489eb

**Pull request merged**